pastas_plugins/pest/solver.py

```diff
@@ -0,0 +1,1159 @@
+        parameters.index = [
+            p.replace("_A", "_g") if p.endswith("_A") else p for p in parameters.index
+        ]
+        parameters.index.name = "parnames"
+        if "constant_d" in parameters.index:
+            if np.isnan(parameters.at["constant_d", "pmin"]):
+                self.ml.set_parameter(
+                    "constant_d", pmin=parameters.at["constant_d", "initial"] - 10.0
+                )
+            if np.isnan(parameters.at["constant_d", "pmax"]):
+                self.ml.set_parameter(
+                    "constant_d", pmax=parameters.at["constant_d", "initial"] + 10.0
+                )
+
+        par_sel = parameters.loc[:, ["optimal"]]
+        par_sel.to_csv(self.model_ws / "parameters_sel.csv")
+        copy_file(self.model_ws / "parameters_sel.csv", self.temp_ws)
+        self.par_sel = par_sel
+
+        # observations
+        observations = self.ml.observations()
+        observations.name = "Observations"
+        observations.to_csv(self.model_ws / "simulation.csv")
+        copy_file(self.model_ws / "simulation.csv", self.temp_ws)
+        self.observations = observations
+
+        # model
+        self.ml.to_file(self.model_ws / "model.pas")
+        copy_file(self.model_ws / "model.pas", self.temp_ws)
+
+        # write run function
+        self.write_run_function()
+
+    def write_pst(self, pst: pyemu.Pst, version: int = 2) -> None:
+        """Write pest control file
+
+        Parameters
+        ----------
+        pst : pyemu.Pst
+            Pyemu pest control file object.
+        version : int, optional
+            Version of the control file, by default version 2
+        """
+        pst.write(self.pf.new_d / "pest.pst", version=version)
+
+    def setup_files(self, version: int = 2):
+        """Setup PEST file structure for optimization
+
+        Parameters
+        ----------
+        version : int, optional
+            Version of the control file, by default version 2
+        """
+
+        # parameters
+        self.pf.add_parameters(
+            self.model_ws / "parameters_sel.csv",
+            index_cols=[self.par_sel.index.name],
+            use_cols=self.par_sel.columns.to_list(),
+            par_type="grid",
+            par_style="direct",
+            transform="none",
+            # pargp=self.par_sel.columns.to_list(),
+            # par_name_base=self.par_sel.columns.to_list(), #[x.split("_")[0] for x in self.par_sel.columns],
+            # lower_bound=self.ml.parameters.loc[self.vary, "pmin"].values.tolist(),
+            # upper_bound=self.ml.parameters.loc[self.vary, "pmax"].values.tolist(),
+            # ult_lbound = self.ml.parameters.loc[self.vary, ["pmin"]].transpose().values.tolist(),
+            # ult_ubound = self.ml.parameters.loc[self.vary, ["pmax"]].transpose().values.tolist(),
+        )
+
+        # observations and simulation
+        self.pf.add_observations(
+            "simulation.csv",
```

Comment:
Why not observations.csv?

Reply:
PEST linguistics where simulated values are also called observations :)

Reply:
Maybe a comment to that effect (or some nomenclature for PASTAS --> PESTAS) would be helpful for future people contributing.